odl/test/space/npy_tensors_test.py

```diff
@@ -0,0 +1,1545 @@
+    NumpyTensorSpace((3, 4), dtype=complex)
+    NumpyTensorSpace((3, 4), dtype=complex, order='F')
+    NumpyTensorSpace((3, 4), dtype=complex, exponent=1.0)
+    NumpyTensorSpace((3, 4), dtype=complex, exponent=float('inf'))
+
+    # Alternative constructor
+    odl.tensor_space((3, 4))
+    odl.tensor_space((3, 4), dtype=int)
+    odl.tensor_space((3, 4), order='F')
+    odl.tensor_space((3, 4), exponent=1.0)
+
+    # Only works with scalar data types
+    with pytest.raises(ValueError):
+        NumpyTensorSpace((3, 4), dtype='S1')
+
+    # Constructors for real spaces
+    odl.rn((3, 4))
+    odl.rn((3, 4), dtype='float32')
+    odl.rn(3)
+    odl.rn(3, dtype='float32')
+
+    # Works only for real data types
+    with pytest.raises(ValueError):
+        odl.rn((3, 4), complex)
+    with pytest.raises(ValueError):
+        odl.rn(3, int)
+    with pytest.raises(ValueError):
+        odl.rn(3, 'S1')
+
+    # Constructors for complex spaces
+    odl.cn((3, 4))
+    odl.cn((3, 4), dtype='complex64')
+    odl.cn(3)
+    odl.cn(3, dtype='complex64')
+
+    # Works only for complex data types
+    with pytest.raises(ValueError):
+        odl.cn((3, 4), float)
+    with pytest.raises(ValueError):
+        odl.cn(3, 'S1')
+
+    # Backported int from future fails (not recognized by numpy.dtype())
+    # (Python 2 only)
+    from builtins import int as future_int
+    if PYTHON2:
+        with pytest.raises(ValueError):
+            NumpyTensorSpace((3, 4), future_int)
+
+    # Init with weights or custom space functions
+    weight_const = 1.5
+    weight_arr = _pos_array(odl.rn((3, 4), float))
+
+    odl.rn((3, 4), weighting=weight_const)
+    odl.rn((3, 4), weighting=weight_arr)
+
+
+def test_init_tspace_weighting(weight, exponent):
+    """Test if weightings during init give the correct weighting classes."""
+    space = odl.tensor_space((3, 4), weighting=weight, exponent=exponent)
+    if isinstance(weight, np.ndarray):
+        weighting = NumpyTensorSpaceArrayWeighting(weight, exponent=exponent)
+    elif weight is None:
+        weighting = NumpyTensorSpaceNoWeighting(exponent=exponent)
+    else:
+        weighting = NumpyTensorSpaceConstWeighting(weight, exponent=exponent)
+
+    assert space.weighting == weighting
+
+    # Using the class directly
+    space = odl.tensor_space((3, 4), weighting=weighting, exponent=exponent)
+    assert space.weighting is weighting
+
+    # Errors for bad input
+    with pytest.raises(ValueError):
+        odl.tensor_space((3, 4), weighting=np.ones([2, 4]))  # bad size
+
+    with pytest.raises(ValueError):
+        odl.tensor_space((3, 4), weighting=1j * np.ones([2, 4]))  # bad dtype
+
+    with pytest.raises(TypeError):
+        odl.tensor_space((3, 4), weighting=1j)  # raised by float() conversion
+
+
+def test_properties():
+    """Test that the space and element properties are as expected."""
+    tspace = odl.tensor_space((3, 4), dtype=complex, exponent=1, weighting=2)
```

Comment:
overall I'd just call it `space` instead of `tspace`. This is a general comment.